### ISSUE TITLE
Enable boolean flags parsing in CLI 

### ISSUE THREAD

**Issue by pvdlg** [xojs/xo#299]
Currently the `semicolon` and `prettier` parsing by `meow` had to be disabled, as if they are not used by the user, `meow` set them to false and they override the configuration file.

So setting `"semicolon": true` or `"prettier": true` in the `package.json` is always overwritten and set to `false`.

Investigating it's related to to [minimist](https://github.com/substack/minimist) which is used internally by `meow`. See https://github.com/substack/minimist/issues/94.
 
`minimist` doesn't seems to be maintained anymore and didn't received any commits for more than 2 years.

In order to solve this issue I see a couple solutions:
- Replace `meow` by another arg parsing solution. In my experience [yargs](https://github.com/yargs/yargs) seems to be the most complete, reliable and maintained solution. [Caporal.js](https://github.com/mattallty/Caporal.js) seems promising as well.
- Replace `minimist` inside `meow` by another solution. [yargs-parser](https://github.com/yargs/yargs-parser) seems a good candidate.

@sindresorhus as you are also the author of `meow`, how would you prefer to address that issue?

**Comment by kevva** [comment]
@pvdlg, I would be open to add support for it in https://github.com/kevva/argvments.

**Comment by pvdlg** [comment]
As https://github.com/kevva/argvments uses `minimist` internally, wouldn't the problem be the same?

**Comment by pvdlg** [comment]
Even if we can always find workaround `minimist`, maybe that's a good occasion to stop relying on it as it's not maintained anymore.

**Comment by kevva** [comment]
I'll just handle it after I've parsed them with `minimist`.

> Even if we can always find workaround minimist, maybe that's a good occasion to stop relying on it as it's not maintained anymore.

Maybe...

**Comment by pvdlg** [comment]
A workaround would be to set the `default` property of `boolean` flags to `null` and delete the `null` prop from the result option object.

```js
prettier: {
	type: 'boolean',
	default: null
}
```

then
```js
const {pickBy} = require('lodash');
const opts = pickBy(cli.flags, value => value !== null))
```

Still we can have discuss, decide if we want to move away from `minimist` based solution or not, and what would be the best solution.

**Comment by pvdlg** [comment]
I tried to implement the workaround mentioned above in `meow` but unfortunately it doesn't work due to another `minimist` bug that seems related to `alias` and `default` props.

```js
minimist(['-f'], {boolean: ['foo'], alias: {f: 'foo'}, default: {foo: null}});
// => {_: [], foo: true, f: [null, true]}
// Should be {_: [], foo: true, f:  true}

minimist(['--foo'], {boolean: ['foo'], alias: {f: 'foo'}, default: {foo: null}});
// => {_: [], foo: true, f: [null, true]}
// Should be {_: [], foo: true, f:  true}
```

For info:
```js
yargsParser(['--foo'], {boolean: ['foo'], alias: {f: 'foo'}, default:{foo: null}});
// => {_: [], foo: true, f: true}

yargsParser(['-f'], {boolean: ['foo'], alias: {f: 'foo'}, default:{foo: null}});
// => {_: [], foo: true, f: true}
```

**Comment by pvdlg** [comment]
I opened sindresorhus/meow#77 as a potential, that relies on [yargs-parser](https://github.com/yargs/yargs-parser).

I can also quickly create a PR for XO that would use [yargs](https://github.com/yargs/yargs) directly, to see how the code would look like.

**Comment by sindresorhus** [comment]
> I opened sindresorhus/meow#77 as a potential, that relies on yargs-parser.

I like it.

> I can also quickly create a PR for XO that would use yargs directly, to see how the code would look like.

Sure. Meow was never meant for huge CLIs like XO. I just made it to abstract common things in the small CLIs I make. Yargs is much better suited for our needs in XO.

I still want the sindresorhus/meow#77 change in Meow though.

**Comment by pvdlg** [comment]
Ok, so you have the PR opened in `meow`. We can improve it in another PR to add the other `yargs-parser` options (besides `alias`, `string`, `default`, `boolean`) if you'd like.

I'll open a PR for XO with `yargs` tonight.

**Comment by sindresorhus** [comment]
> We can improve it in another PR to add the other yargs-parser options (besides alias, string, default, boolean) if you'd like.

That would be amazing.

**Comment by pvdlg** [comment]
With sindresorhus/meow#77 we can just use `booleanDefault: undefined` to solve this issue.
We can still move to `yargs`, but we would have to filter out `undefined` values in the resulting options object due to yargs/yargs#1071.
So unless there is `yargs` feature we'd like to have it's not necessary to migrate.

**Comment by pvdlg** [comment]
With sindresorhus/meow#77 this issue can be solved by setting the boolean flag default to `undefined` or by using `booleanDefault: undefined`.

We can still migrate to `yargs` if we need some of its features, but we will have to filter out `undefined` values due to yargs/yargs#1071.

The possible `yargs` features that could be useful:
- Auto-generation of the help message => slightly simpler code for us
- Stricter validation => Display an error if the user enter un unknown option
- Easier way to set multi-value flags with `array` type => can be either `xo --env browser jquery` or `xo --env browser --env jquery` while currently only the second forms works
- Additional type like `number` or `count` but we don't need them at the moment

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/meow/pull/77 is now available in Meow 5.0.0.
